### PR TITLE
Added mutexes on maps

### LINKED ISSUE #44: Issue with pacman files database
Pacoloco crashed after I tried to search for which package had a specific binary. The specific command I used was:

`sudo pacman -Fy /usr/bin/host`

The log of Pacoloco is as follows:

```
Jan 10 19:14:20 alpha systemd[1]: Started Pacoloco caching proxy server.
Jan 10 19:14:20 alpha pacoloco[2173366]: pacoloco.go:102: Reading config file from /etc/pacoloco.yaml
Jan 10 19:14:20 alpha pacoloco[2173366]: prefetch.go:34: The prefetching routine will be run on 2022-01-11 03:00:00.000036132 -0800 PST m=+27939.112777436
Jan 10 19:14:20 alpha pacoloco[2173366]: prefetch_db.go:78: Creating /var/cache/pacoloco/sqlite-pkg-cache.db
Jan 10 19:14:20 alpha pacoloco[2173366]: pacoloco.go:129: Starting server at port 9129
Jan 10 19:16:30 alpha pacoloco[2173366]: pacoloco.go:446: downloading https://arch.mirror.constant.com//community/os/x86_64/community.db
Jan 10 19:16:30 alpha pacoloco[2173366]: pacoloco.go:446: downloading https://arch.mirror.constant.com//extra/os/x86_64/extra.db
Jan 10 19:16:30 alpha pacoloco[2173366]: pacoloco.go:446: downloading https://arch.mirror.constant.com//core/os/x86_64/core.db
Jan 10 19:17:26 alpha pacoloco[2173366]: pacoloco.go:138: open /var/cache/pacoloco/pkgs/archlinux/extra.db.sig: no such file or directory
Jan 10 19:17:26 alpha pacoloco[2173366]: pacoloco.go:138: open /var/cache/pacoloco/pkgs/archlinux/community.db.sig: no such file or directory
Jan 10 19:17:26 alpha pacoloco[2173366]: pacoloco.go:138: open /var/cache/pacoloco/pkgs/archlinux/core.db.sig: no such file or directory
Jan 10 19:54:04 alpha pacoloco[2173366]: pacoloco.go:446: downloading https://arch.mirror.constant.com//extra/os/x86_64/extra.files
Jan 10 19:54:04 alpha pacoloco[2173366]: pacoloco.go:446: downloading https://arch.mirror.constant.com//core/os/x86_64/core.files
Jan 10 19:54:04 alpha pacoloco[2173366]: pacoloco.go:446: downloading https://arch.mirror.constant.com//community/os/x86_64/community.files
Jan 10 19:54:04 alpha pacoloco[2173366]: prefetch.go:78: error: package with name 'core.files' cannot be prefetched cause it doesn't follow the package name formatting regex
Jan 10 19:54:05 alpha pacoloco[2173366]: prefetch.go:78: error: package with name 'extra.files' cannot be prefetched cause it doesn't follow the package name formatting regex
Jan 10 19:54:05 alpha pacoloco[2173366]: prefetch.go:78: error: package with name 'community.files' cannot be prefetched cause it doesn't follow the package name formatting regex
Jan 10 19:54:22 alpha pacoloco[2173366]: prefetch.go:78: error: package with name 'extra.files' cannot be prefetched cause it doesn't follow the package name formatting regex
Jan 10 19:54:22 alpha pacoloco[2173366]: pacoloco.go:486: serving cached file /var/cache/pacoloco/pkgs/archlinux/extra.files
Jan 10 19:54:22 alpha pacoloco[2173366]: prefetch.go:78: error: package with name 'extra.files' cannot be prefetched cause it doesn't follow the package name formatting regex
Jan 10 19:54:22 alpha pacoloco[2173366]: prefetch.go:78: error: package with name 'community.files' cannot be prefetched cause it doesn't follow the package name formatting regex
Jan 10 19:54:22 alpha pacoloco[2173366]: pacoloco.go:486: serving cached file /var/cache/pacoloco/pkgs/archlinux/community.files
Jan 10 19:54:22 alpha pacoloco[2173366]: prefetch.go:78: error: package with name 'community.files' cannot be prefetched cause it doesn't follow the package name formatting regex
Jan 10 19:54:22 alpha pacoloco[2173366]: prefetch.go:78: error: package with name 'core.files' cannot be prefetched cause it doesn't follow the package name formatting regex
Jan 10 19:54:22 alpha pacoloco[2173366]: pacoloco.go:486: serving cached file /var/cache/pacoloco/pkgs/archlinux/core.files
Jan 10 19:54:22 alpha pacoloco[2173366]: prefetch.go:78: error: package with name 'core.files' cannot be prefetched cause it doesn't follow the package name formatting regex
Jan 10 19:54:40 alpha pacoloco[2173366]: fatal error: concurrent map writes
Jan 10 19:54:40 alpha pacoloco[2173366]: goroutine 1197 [running]:
Jan 10 19:54:40 alpha pacoloco[2173366]: runtime.throw({0x562fa2341285, 0x562fa1f6b3d4})
Jan 10 19:54:40 alpha pacoloco[2173366]:         runtime/panic.go:1198 +0x71 fp=0xc00004d620 sp=0xc00004d5f0 pc=0x562fa1ee5d51
Jan 10 19:54:40 alpha pacoloco[2173366]: runtime.mapassign_faststr(0x562fa2441100, 0xc0001f8ae0, {0xc0000287f8, 0x18})
...
```

Full log here: https://gist.github.com/krathalan/e1147dd4ab036f63847e2a9363843280

### PR DESCRIPTION
**Pull request by Focshole**
Added mutexes on all maps that may happen to be used concurrently. This should fix #44.

### CODE REVIEW

File: mirrorlist.go [hunk 1]
```diff
@@ -10,26 +10,36 @@ import (
 )
 
 func updateMirrorlists() {
+	configReposMutex.RLock()
```
Comment (Focshole) [inline mirrorlist.go:13]: I wrote this but I don't like it at all, all the alternatives I had thought about look bad to me.
 - Switching from config.Repos maps to config.Repos syncMaps on config would have caused more harm than good, it doesn't look like [the go yaml package](https://pkg.go.dev/gopkg.in/yaml.v3) would work as expected on those. Also, this is the only for range on which configs are being used.
- The alternative (horrible) approach would be to hard copy the config.Repos map, work on that, then soft copy it back

**Comment by anatol** [comment]
Another (and better) way to handle concurency with maps is to use `sync.Map` struct instead https://pkg.go.dev/sync#Map

**Comment by Focshole** [comment]
> Another (and better) way to handle concurrency with maps is to use `sync.Map` struct instead https://pkg.go.dev/sync#Map

Will they work with the yaml package?
From what I have looked so far it doesn't look like they do, given that sync.Map also have different ways to access values (there is no indexing operator `[]`, only Get & Set.
After inspecting (not accurately) the `yaml.Unmarshal` function [here](https://github.com/go-yaml/yaml/blob/v3/yaml.go) it looks like that there is no implementation for that sync.Maps.

**Comment by Focshole** [comment]
I would not recommend switching to `sync.Map` because it would require to add some preprocessing layers (which would require to copy and keep coherency between structures) for the `config`. But, if you prefer this second approach, I'll look into it. Let me know what you prefer to do.

**Comment by anatol** [comment]
Returning back to this change. @Focshole I am fine with either explicit mutex or syncmap, whatever works better.

@Jabbermuggel could you please review this change as well? If it is ok I would like to merge it.

**Comment by Focshole** [comment]
> Returning back to this change. @Focshole I am fine with either explicit mutex or syncmap, whatever works better.
> 
> @Jabbermuggel could you please review this change as well? If it is ok I would like to merge it.

This fix is not ok, I have to work a bit on it, there may still be concurrency issues. I have an idea on how to fix them but little time to do so. I'll update you about it.
The issue is around that for cycle which I didn't like much, I have to create something like a snapshot of all the keys in it and then loop over the snapshot to avoid concurrency issues.

**Comment by Focshole** [comment]
I'll close it now, I'll have to rework some stuff on it and right now I don't have enough free time :(

**Pull request closed without merging**